python-client/giskard/core/model.py

```diff
@@ -0,0 +1,229 @@
+
+
+class Model:
+    meta: ModelMeta
+    clf: PyFuncModel
+    data_preparation_function: any
+
+    def __init__(self,
+                 clf,
+                 model_type: Union[SupportedModelTypes, str],
+                 name: str = None,
+                 data_preparation_function=None,
+                 feature_names=None,
+                 classification_threshold=0.5,
+                 classification_labels=None) -> None:
+        self.clf = clf
+        self.data_preparation_function = data_preparation_function
+
+        if type(model_type) == str:
+            try:
+                model_type = SupportedModelTypes(model_type)
+            except ValueError as e:
+                available_values = {i.value for i in SupportedModelTypes}
+                raise ValueError(
+                    f'Invalid model type value "{model_type}". Available values are: {available_values}') from e
+
+        self.meta = ModelMeta(
+            name=name if name is not None else self.__class__.__name__,
+            model_type=model_type,
+            feature_names=feature_names,
+            classification_labels=classification_labels,
+            classification_threshold=classification_threshold
+        )
+
+    @property
+    def is_classification(self):
+        return self.meta.model_type == SupportedModelTypes.CLASSIFICATION
+
+    @property
+    def is_regression(self):
+        return self.meta.model_type == SupportedModelTypes.REGRESSION
+
+    def save(self, client: GiskardClient, project_key, validate_ds=None):
+        from giskard.core.model_validation import validate_model
+
+        validate_model(model=self, validate_ds=validate_ds)
+        with tempfile.TemporaryDirectory(prefix="giskard-model-") as f:
+            info = self.save_to_local_dir(f)
+            self.save_data_preparation_funciton(f)
+            if client is not None:
+                client.log_artifacts(f, posixpath.join(project_key, "models", info.model_uuid))
+                client.save_model_meta(project_key,
+                                       info.model_uuid,
+                                       self.meta,
+                                       info.flavors['python_function']['python_version'],
+                                       get_size(f))
+        return info.model_uuid
+
+    def save_data_preparation_funciton(self, path):
+        if self.data_preparation_function:
+            with open(Path(path) / "giskard-data-prep.pkl", 'wb') as f:
+                cloudpickle.dump(self.data_preparation_function, f, protocol=pickle.DEFAULT_PROTOCOL)
+
+    @classmethod
+    def read_data_preparation_function_from_artifact(cls, local_path: str):
+        local_path = Path(local_path)
+        file_path = local_path / "giskard-data-prep.pkl"
+        if file_path.exists():
+            with open(file_path, 'rb') as f:
+                return cloudpickle.load(f)
+
+    def save_to_local_dir(self, local_path):
+        if self.is_classification:
```

Comment:
I think we discussed that we should split the dumping of the model and meta into two separate functions right?

Reply:
yes, very good point, like this we won't have to reimplement yaml dumping every time